jgrapht-core/src/main/java/org/jgrapht/alg/scoring/ClusteringCoefficient.java

```diff
@@ -0,0 +1,192 @@
+ *
+ * <p>
+ * This implementation also computes the global clustering coefficient as well as the average clustering coefficient.
+ *
+ * <p>
+ * The running time is $O(|V| + |D|^2) where $|V|$ is the number of vertices and $|D|$ is the maximum degree of a
+ * vertex. The space complexity is $O(|V|)$.
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Alexandru Valeanu
+ */
+public class ClusteringCoefficient<V, E> implements VertexScoringAlgorithm<V, Double> {
+
+    /**
+     * Underlying graph
+     */
+    private final Graph<V, E> graph;
+
+    /**
+     * The actual scores
+     */
+    private Map<V, Double> scores;
+
+    /**
+     * Global Clustering Coefficient
+     */
+    private boolean computed = false;
+    private double globalClusteringCoefficient;
+
+    /**
+     * Average Clustering Coefficient
+     */
+    private boolean computedAverage = false;
+    private double averageClusteringCoefficient;
+
+    /**
+     * Construct a new instance
+     *
+     * @param graph the input graph
+     * @throws NullPointerException if {@code graph} is {@code null}
+     */
+    public ClusteringCoefficient(Graph<V, E> graph) {
+        this.graph = Objects.requireNonNull(graph);
+    }
+
+    /**
+     * Computes the global clustering coefficient. The global clustering coefficient $C$ is defined as
+     * $C = 3 \times number\_of\_triangles / number\_of\_triplets$.
+     *
+     * @return the global clustering coefficient
+     */
+    public double getGlobalClusteringCoefficient() {
+        if (!computed){
+            computeGlobalClusteringCoefficient();
+        }
+
+        return globalClusteringCoefficient;
+    }
+
+    /**
+     * Computes the average clustering coefficient. The average clustering coefficient $\={C}$ is defined as
+     * $\={C} = \frac{\sum_{i=1}^{n} C_i}{n}$ where $n$ is the number of vertices.
+     *
+     * Note: the average is $0$ if the graph is empty
+     *
+     * @return the average clustering coefficient
+     */
+    public double getAverageClusteringCoefficient(){
+        if (graph.vertexSet().size() == 0)
+            return 0;
+
+        if (!computedAverage){
+            computeScoreMap();
+            computedAverage = true;
+            averageClusteringCoefficient = 0;
+
+            for (double value: scores.values())
+                averageClusteringCoefficient += value;
+
+            averageClusteringCoefficient /= graph.vertexSet().size();
+        }
+
+        return averageClusteringCoefficient;
+    }
+
+    private void computeGlobalClusteringCoefficient(){
+        computed = true;
+        double numberTriplets = 0;
+
+        for (V v: graph.vertexSet()){
+            if (graph.getType().isUndirected()){
+                numberTriplets += 1.0 * graph.degreeOf(v) * (graph.degreeOf(v) - 1) / 2;
+            }
+            else{
+                numberTriplets += 1.0 * graph.inDegreeOf(v) * graph.outDegreeOf(v);
```

Comment:
Is this valid? For two vertices, 1 and 2, there may be two arcs (1,2) and (2,1). Yet this is not a triplet.